src/v/cloud_topics/reader/placeholder_extent.cc

```diff
@@ -0,0 +1,315 @@
+        auto is_cached_result
+          = result_from_ready_future<errc::cache_read_error>(
+            co_await ss::coroutine::as_future(
+              cache->is_cached(cache_file_name)));
+
+        if (!is_cached_result.has_value()) {
+            co_return is_cached_result.error();
+        }
+
+        switch (is_cached_result.value()) {
+        case cloud_io::cache_element_status::available:
+        case cloud_io::cache_element_status::not_available:
+            status = is_cached_result.value();
+            break;
+        case cloud_io::cache_element_status::in_progress:
+            // Another fiber is trying to put value into the cache.
+            // Wait until the operation is completed but stay within the
+            // time budget.
+            if (rp.abort_source != nullptr) {
+                co_await ss::sleep_abortable(rp.delay, *rp.abort_source);
+            } else {
+                co_await ss::sleep<Clock>(rp.delay);
+            }
+            rp = is_cached_rtc.retry();
+            continue;
+        }
+    }
+
+    if (!rp.is_allowed) {
+        co_return errc::timeout;
+    }
+
+    if (
+      !status.has_value()
+      || status.value() == cloud_io::cache_element_status::available) {
+        auto res = co_await materialize_from_cache(cache_file_name, cache);
+        if (res.has_error()) {
+            co_return res.error();
+        }
+        L0_object->payload = std::move(res.value());
+    } else {
+        auto res = co_await materialize_from_cloud_storage(
+          cache_file_name, bucket, api, cache, rtc);
+        if (res.has_error()) {
+            co_return res.error();
+        }
+        L0_object->payload = std::move(res.value());
+    }
+    co_return hydrated;
+}
+
+template<class Clock>
+ss::future<result<iobuf>>
+basic_placeholder_extent<Clock>::materialize_from_cache(
+  std::filesystem::path cache_file_name,
+  cloud_io::basic_cache_service_api<Clock>* cache) {
+    iobuf result_buf;
+
+    auto buffer_size = config::shard_local_cfg().storage_read_buffer_size();
+    auto read_ahead = config::shard_local_cfg().storage_read_readahead_count();
+    auto io_priority = ss::default_priority_class(); // FIXME
+    auto fut = co_await ss::coroutine::as_future(
+      cache->get(cache_file_name, io_priority, buffer_size, read_ahead));
+    auto sz_stream_result = result_from_ready_future<errc::cache_read_error>(
+      std::move(fut));
+    if (sz_stream_result.has_error()) {
+        co_return sz_stream_result.error();
+    }
+    auto sz_stream = std::move(sz_stream_result.value());
+    if (!sz_stream.has_value()) {
+        co_return errc::cache_read_error;
+    }
+
+    auto target = make_iobuf_ref_output_stream(result_buf);
+    co_await ss::copy(sz_stream->body, target);
+    co_await sz_stream->body.close();
+    co_return result_buf;
+}
+
+template<class Clock>
+ss::future<result<iobuf>>
+basic_placeholder_extent<Clock>::materialize_from_cloud_storage(
```

Comment:
@Lazin what if we make basic_placeholder_extent closer to a dumb object and instead we have some sort of materialization service which does the materialization? Or, just put this methods into the reader. Why we choose to attach so much functionality to the extent class?

Reply:
This is not a final version of this code, don't overthink all this. The goal of the `placeholder` extent is to be used in one place in the read path for a month or two until we will roll out proper read path with reader reuse/caching and centralized control over materialized resources (similar to what we have now in TS). The goal of this code is to unblock further development and be correct.

Reply:
> The goal of this code is to unblock further development and be correct.

yes, but keep things as simple as possible.  i think @nvartolomei is right that it would be better to have a plain object--it would be simpler.

Reply:
@nvartolomei after we merge this, i think you should feel free to come back and do some clean up / refactoring work for stuff like this. i'm planning to do this as well as a forcing function to get into the code in more detail as well.

Reply:
I'll make the change and we will see if it's getting much simpler.